pfunk/_resultsdb.py

```diff
@@ -0,0 +1,162 @@
+#
+# Results DB module.
+#
+# This file is part of Pints Functional Testing.
+#  Copyright (c) 2017-2019, University of Oxford.
+#  For licensing information, see the LICENSE file distributed with the Pints
+#  functional testing software package.
+#
+
+from __future__ import print_function
+from collections import defaultdict
+import sqlite3
+import json
+import time
+import pfunk
+
+
+class ResultsDatabaseSchemaClient(object):
+    primary_columns = ["identifier"]
+    mapped_columns = {"commit": "commit_hashes"}
+    columns = ["name", "date", "status", "python", "pints", "pints_commit", "pfunk_commit", "commit_hashes", "seed",
+               "method"]
+
+    def json_values(self):
+        result = self._connection.execute("select json from test_results where identifier = ?",
+                                          [self._row])
+        json_field = result.fetchone()[0]
+        dictionary = {}
+        if json_field is not None:
+            dictionary = json.loads(json_field)
+        return dictionary
+
+
+class ResultsDatabaseWriter(ResultsDatabaseSchemaClient):
+    """
+    Provides read-write access to a SQLite3 database containing test results.
+    For compatibility with the interfaces supplied by ResultsWriter/ResultsReader,
+    the flat-file equivalents, instances of this class accept a test name and date,
+    and provide access to the values in the row matching those properties only.
+    """
+
+    def __init__(self, filename, test_name, date):
+        self._connection = connect_to_database(filename)
+        self.__ensure_schema()
+        self._filename = filename
+        self._name = test_name
+        self._date = date
+        self.__ensure_row()
+
+    def __ensure_schema(self):
+        query = """ create table if not exists test_results(
```

Comment:
Also want docstrings even for hidden methods (unless they're built-in overrides), just so that we know what the methods are supposed to do (then we can easily check if the code actually does...)